File: python/lombokalgoritma/geometry/cross.py
# LombokAlgoritma — 2-D cross product (SPEC §13.2)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from .types import Point2D


def cross(o: Point2D, a: Point2D, b: Point2D) -> float:
    """(A.x − O.x)·(B.y − O.y) − (A.y − O.y)·(B.x − O.x): > 0 counter-clockwise turn."""
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])

File: typescript/src/geometry/cross.ts
// LombokAlgoritma — 2-D cross product
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import type { Point2D } from './types.js';

/**
 * z-component of (A − O) × (B − O): > 0 when O→A→B turns counter-clockwise, < 0 clockwise,
 * 0 when collinear. Evaluated exactly as `(A.x − O.x)·(B.y − O.y) − (A.y − O.y)·(B.x − O.x)`
 * (SPEC §7.1).
 */
export function cross(O: Point2D, A: Point2D, B: Point2D): number {
  return (A.x - O.x) * (B.y - O.y) - (A.y - O.y) * (B.x - O.x);
}
